### PR TITLE
fix SELU gain

### LINKED ISSUE #24991: [Feature request] Add support for selu activation to calculate_gain function.
## 🚀 Feature

Add support for selu activation to calculate_gain function.

https://pytorch.org/docs/stable/_modules/torch/nn/init.html#calculate_gain

cc @albanD @mruberry @SsnL

### PR DESCRIPTION
**Pull request by mrTsjolder**
Fixes #24991 

This issue has been wrongly tackled by PR ##50664 in the sense that the [original paper](https://proceedings.neurips.cc/paper/2017/hash/5d44ee6f2c3f71b73125876103c8f6c4-Abstract.html) explicitly requires the variance of initial weights to be `1/fan_in`.

### CODE REVIEW

**Comment by jbschlosser** [comment]
Reading the [paper](https://arxiv.org/pdf/1706.02515.pdf), the authors clearly propose variance of `1 / fan_in` as described by @mrTsjolder: "For the weight initialization, we propose `ω = 0` and `τ = 1` for all units in the higher layer." (τ being defined as "n times the second moment", where `n` is `fan_in`).

Their proof that the mapping between layers has a fixed point with `(0, 1)` mean / variance that is stable and attracting assumes the above initialization. They also prove that initializations "close to" the above will result in stable and attracting fixed points "close to" `(0, 1)` mean / variance, with an assumed range `τ ∈ [0.95, 1.1]`.

Since the chosen empirical gain value of `0.75` corresponds with τ outside of this range, there's no reason (from the paper at least) to believe that this empirical value will result in a stable and attracting fixed point near `(0, 1)` mean / variance (i.e. self-normalizing behavior).

TL;DR: I agree the gain should be fixed by setting it to `1`.

**Comment by t-vi** [comment]
The process that got to 0.75 was
1. I have read the Klambauer et al. paper which suggests 1,
2. I tried the stability with 1 and found it lacking for an initialization, 
3. I tried 0.75 and it worked better.
4. Some people agreed and put 0.75 into PyTorch.

@mrTsjolder and @jbschlosser argue that we should have stopped at 1.

The experiment was given in the PR introducing (credit to jpeg729)
```python
import torch
import torch.nn.functional as F
import sys

a = torch.randn(1000,1000, requires_grad=True)
b = a
print (f"in: {a.std().item():.4f}")
for i in range(100):
    l = torch.nn.Linear(1000,1000, bias=False)
    torch.nn.init.xavier_normal_(l.weight, torch.nn.init.calculate_gain("selu"))
    b = getattr(F, 'selu')(l(b))
    if i % 10 == 0:
        print (f"out: {b.std().item():.4f}", end=" ")
        a.grad = None
        b.sum().backward(retain_graph=True)
        print (f"grad: {a.grad.abs().mean().item():.4f}")
```

The objection that it is fan in vs. fan out seems to be rather unimpressive when all activations are 1000 elements.

TLDR: The reason to put in 0.75 was not *we were too stupid to read the paper* but had to do with actual behaviour of actual initializations.

**Comment by jbschlosser** [comment]
> TLDR: The reason to put in 0.75 was not _we were too stupid to read the paper_ but had to do with actual behaviour of actual initializations.

Haha, I certainly didn't mean to imply this! On the contrary, I appreciate the extra effort put in to empirically determine a well-performing gain value. I'm just worried about the lack of theoretical backing for 0.75 wrt the self-normalizing behavior.

I'd argue that the default should be 1, with the option to override it. This is currently possible for Xavier init, but Kaiming / He init doesn't allow custom gain values iirc. Maybe it makes sense to open a new issue for optional configurable gain with Kaiming / He init?

**Comment by t-vi** [comment]
How about we leave it at 0.75 instead?
[Newer research]((https://arxiv.org/abs/1712.05577)) appears to indicate that SELU with the Klambauer parameters is subject to gradient explosion, which is the same thing that we observed and which choosing 0.75 over 1 mitigates.

Again, we have actual code that demonstrates advantages of 0.75 over 1 and no code demonstrating advantages of 1 over 0.75.

**Comment by albanD** [comment]
I am not an expert on the SELU question but I think there is one more thing to take into consideration, which is the heavily BC-breaking nature of this change. I don't think we want to change the behavior silently here, so that would mean that this will take 3+ releases to actually change IF we decide to do it.
- Add new gain optional argument with default value 0.75
- Deprecate the default value for the gain
- Remove the default value for the gain
- Add back 1. as the default value

Given that, an alternative could be to stop at the first step and add explicit doc as we have reasons to believe that this is better than what the original paper was suggesting:
We would add a gain optional argument with default value of 0.75 and add a warning in the doc explaining that PyTorch uses a gain of 0.75 for historical reasons since it leads to better stability in practice (linking to most recent research). And if the user wants to reproduce exactly the behavior from the original paper, they should set that value to 1.

**Comment by jbschlosser** [comment]
> I am not an expert on the SELU question but I think there is one more thing to take into consideration, which is the heavily BC-breaking nature of this change. I don't think we want to change the behavior silently here, so that would mean that this will take 3+ releases to actually change IF we decide to do it.
> 
> * Add new gain optional argument with default value 0.75
> * Deprecate the default value for the gain
> * Remove the default value for the gain
> * Add back 1. as the default value

I'd much rather keep 0.75 as default than go through that process :P

> We would add a gain optional argument with default value of 0.75 and add a warning in the doc explaining that PyTorch uses a gain of 0.75 for historical reasons since it leads to better stability in practice (linking to most recent research). And if the user wants to reproduce exactly the behavior from the original paper, they should set that value to 1.

Slight modification here: the default gain arg would be for the `kaiming_uniform_` and `kaiming_normal_` functions in `torch/nn/init.py`. They currently use `calculate_gain` with the nonlinearity type to determine gain.

I propose a new `gain=None` arg for these functions that defaults to using `calculate_gain`, but uses the specified gain value if provided. Right now, the user would have to re-implement these functions to use a different gain value.

Note that the `xavier_uniform_` and `xavier_normal_` functions in `torch/nn/init.py` already have configurable gain, so there's some precedent for that.

**Comment by mrTsjolder** [comment]
> > TLDR: The reason to put in 0.75 was not _we were too stupid to read the paper_ but had to do with actual behaviour of actual initializations.
> 
> Haha, I certainly didn't mean to imply this! On the contrary, I appreciate the extra effort put in to empirically determine a well-performing gain value. I'm just worried about the lack of theoretical backing for 0.75 wrt the self-normalizing behavior.

Same here, there is just no proper explanation for why to take 0.75 (apart from the limited empirical validation in the provided script).

> I'd argue that the default should be 1, with the option to override it. This is currently possible for Xavier init, but Kaiming / He init doesn't allow custom gain values iirc. Maybe it makes sense to open a new issue for optional configurable gain with Kaiming / He init?

The default should definitely be 1. After all, people who aim to implement self-normalising networks, would rely on the default initialisation to correspond to the network presented in the original paper (at least that is what I would expect). Another solution would be to emphasise (e.g. in the docs) that `nonlinearity='linear'` should be used to implement SNNs (which is the solution I have been using thus far), but I would argue that is very counter-intuitive.

Maybe it makes sense to also mention that SELU is just a scaled version of the ELU activation function. More specifically `SELU(x) = 1.05 * ELU(x, alpha=1.67)`. The main goal of ELU was to push the mean towards zero, whereas the SELU was introduced to induce a fixed point (i.e. keep data normalised throughout the network). If the normalisation during the forward pass is irrelevant, it would make more sense to use ELU instead of SELU. Therefore, it makes no sense to have a default initialisation for SELUs that breaks the normalisation properties of SNNs, since you would simply be using a scaled variant of the ELU activation function again that is **not** self-normalising.

Since the goal of ELUs is to provide a zero-mean distribution, it might be less sensitive to the initial parameters. The [original paper](https://arxiv.org/abs/1511.07289) mentions that they use He-initialisation (i.e. `gain = 2 ** .5), but this gives pretty bad results (cf. modified script from @t-vi ):
```python
import torch
import torch.nn.functional as F
import sys
 
a = torch.randn(1000,1000, requires_grad=True)
b = a
print (f"in: {a.mean().item():.4f}")
for i in range(100):
    l = torch.nn.Linear(1000,1000, bias=False)
    torch.nn.init.xavier_normal_(l.weight, 2 ** .5)
    b = getattr(F, 'elu')(l(b))
    if i % 10 == 0:
        print (f"out: {b.mean().item():.4f}", end=" ")
        a.grad = None
        b.sum().backward(retain_graph=True)
        print (f"grad: {a.grad.abs().mean().item():.4f}")
``` 
Note that the mean is not conserved at all:
```
in: -0.0001
out: 0.2786 grad: 0.9804
out: 0.6048 grad: 3.1093
out: 0.9645 grad: 6.6008
out: 1.3986 grad: 12.7863
out: 1.4706 grad: 18.6828
out: 1.5861 grad: 29.4972
out: 1.9648 grad: 42.0425
out: 1.8034 grad: 54.9726
out: 2.1447 grad: 85.2120
out: 1.9795 grad: 118.8753
```
Therefore, it might make more sense to use the empirical gain (which should be approximately `1.15` with `alpha=1.0`) for ELU to get good forward and backward dynamics and keep SELU according to the paper (so that it properly normalises data).  

> How about we leave it at 0.75 instead?
> [Newer research](https://arxiv.org/abs/1712.05577) appears to indicate that SELU with the Klambauer parameters is subject to gradient explosion, which is the same thing that we observed and which choosing 0.75 over 1 mitigates.

The main problem with normalisation layers is that they work well in practice, but there it is not clear why they improve learning. I am aware of this research, and this simply states that the claim that normalisation solves the vanishing gradient problem is false. This does not mean that normalisation layers (as e.g. SNN layers) should be discarded. SNNs also seem to work well in practice (despite possibly exploding gradients).

> Again, we have actual code that demonstrates advantages of 0.75 over 1 and no code demonstrating advantages of 1 over 0.75.

The code merely demonstrates that a gain of 0.75 does not increase the variance as much as a gain of 1 for rectangular layers ([this master thesis](https://epub.jku.at/obvulihs/content/titleinfo/2344947) discusses the effect of architecture on the backward dynamics in SNNs). This code does not illustrate any form of performance improvements. Given that the normalisation is broken with this gain, it could even be that performance is much worse with the gain (despite better gradient dynamics). Therefore, I would argue that this code demonstrates a single advantage, but might hide possible disadvantages, and therefore should not be taken as the whole truth.

> I am not an expert on the SELU question but I think there is one more thing to take into consideration, which is the heavily BC-breaking nature of this change. I don't think we want to change the behavior silently here, so that would mean that this will take 3+ releases to actually change IF we decide to do it.

In my humble opinion, there is little choice here. The alternative to this BC-breaking change would be a warning in the docs that states that for SNNs `nonlinearity='linear'` should be used, rather than `nonlinearity='selu'`.

TL;DR: the gain for SELU must be 1! An empirical gain might make sense for ELU.

**Comment by mrTsjolder** [comment]
@albanD @jbschlosser Comments/remarks/updates on further steps?

**Comment by jbschlosser** [comment]
@mrTsjolder I'm proposing in the linked issue above that the `nonlinearity` arg be replaced with a `gain` arg to improve the non-intuitive UX of specifying e.g. `nonlinearity='linear'` to use a gain value outside of the nonlinearity-based defaults.

While I personally agree that the ideal default for `selu` would have been `1` originally, at this point, I'd prefer to maintain BC for users that may depend on the empirically-determined `0.75` default value.

If this isn't satisfactory for you, I encourage you to open a new issue for changing selu's default to 1 and link it to the discussion on this PR. If the conversation there gets enough traction from other users who agree, we can certainly revisit the possibility of changing it.
